streaming/aggregate/avg.go

```diff
@@ -0,0 +1,109 @@
+package aggregate
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+type Average struct {
+	underlyingSum   *Sum
+	underlyingCount *Count
+}
+
+func NewAverageAggregate() *Average {
+	return &Average{
+		underlyingSum:   NewSumAggregate(),
+		underlyingCount: NewCountAggregate(),
+	}
+}
+
+func (agg *Average) AddValue(ctx context.Context, tx storage.StateTransaction, value octosql.Value) error {
+	valueType := value.GetType()
+
+	if !isAppropriateType(valueType) {
+		return errors.Errorf("type of value passed (%s) isn't appropriate for calculating average", valueType)
+	}
+
+	err := agg.underlyingSum.AddValue(ctx, tx, value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't add value to sum for avg")
+	}
+
+	err = agg.underlyingCount.AddValue(ctx, tx, value)
+	if err != nil {
+		err2 := agg.underlyingSum.RetractValue(ctx, tx, value) // TODO - is this necessary
+		if err2 != nil {
+			return errors.Wrap(err2, "couldn't retract value from sum for avg after error occured in count update")
+		}
+
+		return errors.Wrap(err, "couldn't add element to elements count for avg")
+	}
+
+	return nil
+}
+
+func isAppropriateType(valueType octosql.Type) bool {
+	return valueType == octosql.TypeInt || valueType == octosql.TypeFloat
```

Comment:
I think we can also calculate an average of Durations

Reply:
maybe, it was quite not obvious if we do, what do you think @cube2222 ?

Reply:
Yes, agreed.